From a5da22c9a726c48ec226f82455fb680e1a75d21e Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:36 -0400
Subject: [fs] btrfs: Add nodiscard mount option

Message-id: <1407179109-13714-90-git-send-email-sandeen@redhat.com>
Patchwork-id: 87002
O-Subject: [RHEL7.1 PATCH 089/422] btrfs: Add nodiscard mount option.
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Qu Wenruo <quwenruo@cn.fujitsu.com>

commit e07a2ade4426a2cbafae4018aa7b6944bb627a6e
Author: Qu Wenruo <quwenruo@cn.fujitsu.com>
Date:   Mon Jan 6 09:58:27 2014 +0800

    btrfs: Add nodiscard mount option.

    Add nodiscard mount option to disable discard with remount option.

    Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/filesystems/btrfs.txt b/Documentation/filesystems/btrfs.txt
index 4c48138..1154901 100644
--- a/Documentation/filesystems/btrfs.txt
+++ b/Documentation/filesystems/btrfs.txt
@@ -98,9 +98,12 @@ Options with (*) are default options and will not show in the mount options.
  can be avoided.  Especialy useful when trying to mount a multi-device
  setup as root.  May be specified multiple times for multiple devices.
 
+  nodiscard(*)
   discard
- Issue frequent commands to let the block device reclaim space freed by
- the filesystem.  This is useful for SSD devices, thinly provisioned
+ Disable/enable discard mount option.
+ Discard issues frequent commands to let the block device reclaim space
+ freed by the filesystem.
+ This is useful for SSD devices, thinly provisioned
  LUNs and virtual machine images, but may have a significant
  performance impact.  (The fstrim command is also available to
  initiate batch trims from userspace).
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 44513f3..e153770 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -323,7 +323,7 @@ enum {
  Opt_no_space_cache, Opt_recovery, Opt_skip_balance,
  Opt_check_integrity, Opt_check_integrity_including_extent_data,
  Opt_check_integrity_print_mask, Opt_fatal_errors, Opt_rescan_uuid_tree,
- Opt_commit_interval, Opt_barrier, Opt_nodefrag,
+ Opt_commit_interval, Opt_barrier, Opt_nodefrag, Opt_nodiscard,
  Opt_err,
 };
 
@@ -351,6 +351,7 @@ static match_table_t tokens = {
  {Opt_flushoncommit, "flushoncommit"},
  {Opt_ratio, "metadata_ratio=%d"},
  {Opt_discard, "discard"},
+ {Opt_nodiscard, "nodiscard"},
  {Opt_space_cache, "space_cache"},
  {Opt_clear_cache, "clear_cache"},
  {Opt_user_subvol_rm_allowed, "user_subvol_rm_allowed"},
@@ -575,6 +576,9 @@ int btrfs_parse_options(struct btrfs_root *root, char *options)
   case Opt_discard:
    btrfs_set_opt(info->mount_opt, DISCARD);
    break;
+  case Opt_nodiscard:
+   btrfs_clear_opt(info->mount_opt, DISCARD);
+   break;
   case Opt_space_cache:
    btrfs_set_opt(info->mount_opt, SPACE_CACHE);
    break;
-- 
1.7.1